vochain/ist/ist.go

```diff
@@ -0,0 +1,303 @@
+func NewISTC(s *state.State) *Controller {
+	return &Controller{
+		st: s,
+	}
+}
+
+const (
+	// ActionComputeResults computes and schedules the commit of results, for the given election and height.
+	ActionComputeResults = iota
+	// ActionCommitResults schedules the commit of the results to the state.
+	ActionCommitResults
+	// ActionEndProcess sets a process as ended. It schedules ActionComputeResults.
+	ActionEndProcess
+)
+
+// ActionsToString translates the action identifers to its corresponding human friendly string.
+var ActionsToString = map[int]string{
+	ActionComputeResults: "compute-results",
+	ActionCommitResults:  "commit-results",
+	ActionEndProcess:     "end-process",
+}
+
+// Actions is the model used to store the list of IST actions for
+// a specific height into state.
+type Actions struct {
+	Actions map[string]Action
+}
+
+// Action is the model used to store the IST actions into state.
+type Action struct {
+	Action     int
+	ElectionID []byte
+	Attempts   uint32
+}
+
+// encode performs the encoding of the IST action using Gob.
+func (m *Actions) encode() []byte {
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(m); err != nil {
+		// should never happend
+		panic(err)
+	}
+	return buf.Bytes()
+}
+
+// decode performs the decoding of the IST action using Gob.
+func (m *Actions) decode(data []byte) error {
+	return gob.NewDecoder(bytes.NewReader(data)).Decode(m)
+}
+
+// Schedule schedules an IST action to be executed at the given height.
+func (c *Controller) Schedule(height uint32, id []byte, action Action) error {
+	if id == nil {
+		return fmt.Errorf("cannot schedule IST action: nil id")
+	}
+	actions, err := c.Actions(height)
+	if err != nil {
+		return err
+	}
+	// set the IST action
+	actions.Actions[string(id)] = action
+
+	// store the IST actions
+	log.Debugw("schedule IST action", "height", height, "id", fmt.Sprintf("%x", id), "action", ActionsToString[action.Action])
+	return c.storeToNoState(dbIndex(height), actions.encode())
+}
+
+// Actions returns the IST actions scheduled for the given height.
+// If no actions are scheduled, it returns an empty IstActions.
+func (c *Controller) Actions(height uint32) (*Actions, error) {
+	// get the IST actions
+	actions := Actions{}
+	actionsBytes, err := c.retrieveFromNoState(dbIndex(height))
+	if err != nil {
+		if errors.Is(err, db.ErrKeyNotFound) {
+			// no IST actions scheduled for this height, we return empty
+			actions.Actions = make(map[string]Action)
+			return &actions, nil
+		}
+		return nil, fmt.Errorf("cannot get IST actions on height %d: %w", height, err)
+	}
+	// decode the IST actions
+	if err := actions.decode(actionsBytes); err != nil {
+		return nil, fmt.Errorf("could not decode actions: %w", err)
+	}
+	return &actions, nil
+}
+
+// Reschedule reschedules an IST action to the given new height.
+func (c *Controller) Reschedule(id []byte, oldHeight, newHeight uint32) error {
+	if id == nil {
```

Comment:
probably do `if len(id) == 0` instead, otherwise you'll happily accept `Reschedule(make([]byte, 0), ...` which is still an invalid ID